### PR TITLE
fix(ngcc): compute the correct package paths for a target entry-points

### LINKED ISSUE #40352: NGCC failure with libraries of similar names
# 🐞 bug report

### Affected Package
@angular/compiler-cli - ngcc

### Is this a regression?

Pretty sure it worked in older versions.

### Description

If two libraries have the same name, but the second one has the additional suffix, the ngcc will fail.

**Example**

1. library: `demo-lib`
2. library: `demo-lib-shell`

`demo-lib` imports `demo-lib-shell` module and `demo-lib` is imported by the application.

**Additional observation**

To my surprise, the failure only happens if `demo-lib-shell` library is configured after `demo-lib` in `tsconfig.json` paths, as in:

```JSON
{
  "compilerOptions": {
    "paths": {
      "demo-lib": [
        "dist/demo-lib/demo-lib",
        "dist/demo-lib"
      ],
      "demo-lib-shell": [
        "dist/demo-lib-shell/demo-lib-shell",
        "dist/demo-lib-shell"
      ]
    }
  }
}
```

If this configuration is reversed, `ngcc` will not fail.

## 🔬 Minimal Reproduction

Clone: https://github.com/klemenoslaj/ngcc-failure--library-name
Install dependencies: `yarn`
Reproduce the issue: `yarn reproduce` OR `yarn reproduce:docker`

**NOTE**: There is a change that on the first try right after dependency installation this will succeed, but will fail when `yarn reproduce` is executed again. This probably has something to do with the fact that the rest of dependencies from `node_modules` are processed as well first time, but I could be wrong.

**NOTE 2**: This seems to fail consistantly even on the first try in Docker. To that end I have added a docker file and updated the reproduction above accordingly.

## 🔥 Exception or Error
<pre><code>
⠋ Generating browser application bundles...
Compiling demo-lib-shell : es2015 as esm2015

Compiling demo-lib : es2015 as esm2015
✔ Browser application bundle generation complete.

Error: Failed to compile entry-point demo-lib (es2015 as esm2015) due to compilation errors:
dist/demo-lib-shell/lib/demo-lib-shell.module.d.ts:1:22 - error NG6002: Appears in the NgModule.imports of DemoLibModule, but could not be resolved to an NgModule class.

This likely means that the dependency (demo-lib-shell) which declares DemoLibShellModule has not been processed correctly by ngcc.

1 export declare class DemoLibShellModule {
                       ~~~~~~~~~~~~~~~~~~



error Command failed with exit code 1.
</code></pre>


## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 11.0.6
Node: 14.15.0
OS: darwin x64

Angular: 11.0.7
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1100.6
@angular-devkit/build-angular   0.1100.6
@angular-devkit/core            11.0.6
@angular-devkit/schematics      11.0.6
@angular/cli                    11.0.6
@schematics/angular             11.0.6
@schematics/update              0.1100.6
ng-packagr                      11.0.3
rxjs                            6.6.3
typescript                      4.0.5
</code></pre>

### LINKED ISSUE #40357: Library is skipped by Ivy / ngcc compiler
# 🐞 Bug report

### Command (mark with an `x`)

- [ ] new
- [x] build
- [x] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] extract-i18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Is this a regression?

I'm not sure. Upgraded to angular 11 almost together with getting the issue - might be.

### Description

I maintain a set of libraries to work with gRPC from Angular, here is the link: https://github.com/ngx-grpc/ngx-grpc

The lib includes 6 libs (located in `packages`):
- one non-angular (however located between angular apps, just is not present in angular.json) - protoc plugin - `protoc-gen-ng`
- 4 normal angular libs
- 1 angular lib; built using Angular toolset but is used inside a web worker - `worker`.

There is also `examples` directory served as example and, well, as a test reference that all goes fine. This example uses all the libs, and this is the place where the error is happening.

Now I added a module with forRoot / forChild in some of the angular libraries: core (2 modules), grpc-web-client (1 module) and worker-client (1 module). All modules and libs are working fine, all, but the latter one. It is just not processed by `ngcc`. When I start the application I see the following:

```
Compiling @ngx-grpc/common : es2015 as esm2015

Compiling @ngx-grpc/core : es2015 as esm2015

Compiling @ngx-grpc/- removing all references to the `Worker` type () : es2015 as esm2015

Compiling @ngx-grpc/well-known-types : es2015 as esm2015

Compiling @ngx-grpc/worker : es2015 as esm2015

Compiling @ngx-grpc/worker-client : es2015 as esm2015
```

So, the ngcc reaches the library, but the directory `__ivy_ngcc__` is not created in `dist/worker-client`, however for all other modules it is created. The outcome is that the latter library does not work.

What I tried:
- removing all references to the `Worker` type from `worker-client` - did not help
- checking angular config, including angular.json, tsconfigs and ng-packagr  - did not help, it is totally equal to the working lib `grpc-web-client`
- regenerated the lib from scratch - did not help
- removed references to `worker` library from `worker-client` - did not help
- moved code from `worker-client` to  `grpc-web-client` lib. This **made the code work**, meaning that the `grpc-web-client` had no problems with ngcc; and I can use the code in the example and it works just fine. However, I cannot get more out of it; I still need the code in a separate package. This also means that code is valid and it can be compiled by ngcc, but for some reason when it's in its own package it does not.

What is also very weird - I would not be surprised if I had problems with `worker` library; it's built by Angular and used in Worker. And I have a feeling it can be connected to the issue. 

I don't know what / where else to dig. The issue is driving me crazy. It seems unlogical to me and I cannot find the reason. 

Please help me out.

## 🔬 Minimal Reproduction

- Clone https://github.com/ngx-grpc/ngx-grpc
- Switch to develop branch
- `npm ci`
- `npm run build` - builds the libs
- `npm run examples` - produces the error after ngcc compiler

## 🔥 Exception or Error
<pre><code>
    Error: dist/worker-client/lib/grpc-worker-client.module.d.ts:11:22 - error NG6002: Appears in the NgModule.imports of GrpcWorkerClientExampleModule, but could not be resolved to an NgModule class.
    
    This likely means that the dependency (@ngx-grpc/worker-client) which declares GrpcWorkerClientModule has not been processed correctly by ngcc.
    
    11 export declare class GrpcWorkerClientModule {
</code></pre>

## 🌍 Your Environment
<pre><code>
Angular CLI: 11.0.2
Node: 14.5.0
OS: darwin x64

Angular: 11.0.2
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1100.2
@angular-devkit/build-angular   0.1100.2
@angular-devkit/core            11.0.2
@angular-devkit/schematics      11.0.2
@angular/cdk                    11.0.0
@angular/material               11.0.0
@schematics/angular             11.0.2
@schematics/update              0.1100.2
ng-packagr                      11.0.3
rxjs                            6.6.3
typescript                      4.0.5
</code></pre>

**Anything else relevant?**

I'm not sure wheher the issue belongs to @angular/cli, please feel free to locate it properly in case it's necessary

### PR DESCRIPTION
**Pull request by petebacondarwin**
Previously, if an entry-point started with the string of another package that
is included in `paths` mappings, then the base path was incorrectly computed
resulting in the wrong package path for the entry point.

Now we not only check whether the target path "starts with" the base path
but then also whether the target path is actually contained in the base path
from a file-system directory point of view.

Fixes #40352
Fixes #40357

### CODE REVIEW

File: packages/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.ts [hunk 1]
```diff
@@ -147,6 +147,18 @@ export class TargetedEntryPointFinder extends TracingEntryPointFinder {
     return this.computePackagePathFromNearestNodeModules(entryPointPath);
   }
 
+  /**
+   * Compute whether the `test` path is contained within the `base` path.
+   *
+   * Note that this doesn't use a simple `startsWith()` since that would result in a false positive
+   * for `test` paths `a/b/c-x` when the `base` path is `a/b/c`.
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.ts:154]: Should this be like this? It reads a bit strange currently.

```suggestion
   * for `test` path `a/b/c-x` when the `base` path is `a/b/c`.
```

**Review by JoostK — APPROVED** [review]
There's also a typo in the commit message header, which mentions "a target entry-points"

File: packages/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/compiler-cli/ngcc/src/entry_point_finder/targeted_entry_point_finder.ts:154]: Actually I meant it to be:

```suggestion
   * for `test` paths such as `a/b/c-x` when the `base` path is `a/b/c`.
```

**Comment by petebacondarwin** [comment]
Since the commit message needed updating I just amended the commit.

**Pull request closed without merging**